R/selection.R

```diff
@@ -1,53 +1,70 @@
 # ... are additional arguments passed to trait when trait is a function
 getResponse = function(pop,trait,use,simParam=NULL,...){
-  use = tolower(use)
-  if(use=="rand"){
-    return(rnorm(pop@nInd))
-  }
   if(is(trait,"function")){
-    if(use=="gv"){
-      response = trait(pop@gv,...)
-    }else if(use=="ebv"){
-      response = trait(pop@ebv,...)
-    }else if(use=="pheno"){
-      response = trait(pop@pheno,...)
-    }else if(use=="bv"){
-      if(is(pop,"HybridPop")){
-        stop("Use='bv' is not a valid option for HybridPop")
+    if(is.character(use)){
+      use = tolower(use)
+      if(use=="rand"){
+        return(rnorm(pop@nInd))
+      }else if(use=="gv"){
+        response = trait(pop@gv,...)
+      }else if(use=="ebv"){
+        response = trait(pop@ebv,...)
+      }else if(use=="pheno"){
+        response = trait(pop@pheno,...)
+      }else if(use=="bv"){
+        if(is(pop,"HybridPop")){
+          stop("Use='bv' is not a valid option for HybridPop")
+        }
+        response = genParam(pop,simParam=simParam)$bv
+        response = trait(response,...)
+      }else{
+        stop(paste0("Use=",use," is not an option"))
       }
-      response = genParam(pop,simParam=simParam)$bv
-      response = trait(response,...)
+    }else if(is(use,"function")){
+      response = trait(use(pop, ...), ...)
     }else{
-      stop(paste0("Use=",use," is not an option"))
+      stop("use must be a character or a function")
     }
-  }else{
-    if(is.character(trait)){
-      # Suspect trait is a name
+  }else{ # trait is not a function, so must be numeric or character
+    if(is.character(trait)){ # Suspect trait is a name
       take = match(trait, simParam$traitNames)
       if(is.na(take)){
         stop("'",trait,"' did not match any trait names")
       }
       trait = take
     }
-    if(use == "gv"){
-      response = pop@gv[,trait,drop=FALSE]
-    }else if(use=="ebv"){
-      response = pop@ebv[,trait,drop=FALSE]
-    }else if(use=="pheno"){
-      response = pop@pheno[,trait,drop=FALSE]
-    }else if(use=="bv"){
-      if(is(pop,"HybridPop")){
-        stop("Use='bv' is not a valid option for HybridPop")
+    if(is.character(use)){
+      use = tolower(use)
+      if(use=="rand"){
+        return(rnorm(pop@nInd))
+      }else if(use == "gv"){
+        response = pop@gv[,trait,drop=FALSE]
+      }else if(use=="ebv"){
+        response = pop@ebv[,trait,drop=FALSE]
+      }else if(use=="pheno"){
+        response = pop@pheno[,trait,drop=FALSE]
+      }else if(use=="bv"){
+        if(is(pop,"HybridPop")){
+          stop("Use='bv' is not a valid option for HybridPop")
+        }
+        response = genParam(pop,simParam=simParam)$bv[,trait,drop=FALSE]
+      }else{
+        stop(paste0("Use=",use," is not an option"))
       }
-      response = genParam(pop,simParam=simParam)$bv[,trait,drop=FALSE]
+    }else if(is(use,"function")){
+      response = use(pop, trait=trait, ...)
```

Comment:
I think this will be the most common case when `use` is a function. We likely don't need to really pass `trait` to `use()` but for consistency/flexibility I felt it's probably best.